[FLINK-4692] [tableApi] Add tumbling group-windows for batch tables

wuchong opened the pull request:
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

This PR supports tumbling group-window on batch. 

The design doc please refer to [this](https://docs.google.com/document/d/1lzpnNUmNzn9yuCGf1RSjHuHAWm-O_v2in7y90muXI2o/edit).

twalthr commented:
I would like to shepherd this PR.

twalthr commented:
Thanks for working on this @wuchong! The code looks very good. I haven"t reviewed the code entirely but tried to verify if batch and stream queries return the same result. Unfortunately, this is not the case at the moment.

I converted the stream test `org.apache.flink.api.scala.stream.table.AggregationsITCase#testEventTimeTumblingWindow` and your batch test `org.apache.flink.api.scala.batch.table.AggregationsITCase#testEventTimeTumblingGroupWindowOverTime` but the aggregations are different.

Could you have a look at it?

wuchong commented:
Sure, I will look at it today @twalthr .

wuchong commented:
Hi @twalthr , I have tested and the batch and stream queries return the same result in my machine. The `org.apache.flink.api.scala.stream.table.AggregationsITCase#testEventTimeTumblingWindow` and `org.apache.flink.api.scala.batch.table.AggregationsITCase#testEventTimeTumblingGroupWindowOverTime` have different result because of the different input data. If change batch input data to the same stream input data, they will return the same result. Is that the issue you confusing ? 

BTW, I have rebased the code and fixed the conflicts.

twalthr commented:
I will look into it again. Actually I used the same data, but maybe I did a mistake.

twalthr commented:
Here is an example to reproduce the wrong results:

```scala
  @Test
  def testEventTimeTumblingWindowStream(): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    val tEnv = TableEnvironment.getTableEnvironment(env)

    val stream = env
      .fromCollection(data)
      .assignTimestampsAndWatermarks(new TimestampWithEqualWatermark())
    val table = stream.toTable(tEnv, 'long, 'int, 'string)

    val windowedTable = table
      .groupBy('string)
      .window(Tumble over 5.milli on 'rowtime as 'w)
      .select('string, 'int.count, 'int.avg, 'w.start, 'w.end)

    val results = windowedTable.toDataStream[Row].print()
    env.execute()
  }

  @Test
  def testEventTimeTumblingWindowBatch(): Unit = {
    val env = ExecutionEnvironment.getExecutionEnvironment
    val tEnv = TableEnvironment.getTableEnvironment(env)

    val stream = env
      .fromCollection(data)
    val table = stream.toTable(tEnv, 'long, 'int, 'string)

    val windowedTable = table
      .groupBy('string)
      .window(Tumble over 5.milli on 'long as 'w)
      .select('string, 'int.count, 'int.avg, 'w.start, 'w.end)

    val results = windowedTable.toDataSet[Row].print()
  }
```

It should be: 
```
Hello world,1,3,1970-01-01 00:00:00.005,1970-01-01 00:00:00.01
Hello,2,2,1970-01-01 00:00:00.0,1970-01-01 00:00:00.005
Hello world,1,3,1970-01-01 00:00:00.015,1970-01-01 00:00:00.02
Hi,1,1,1970-01-01 00:00:00.0,1970-01-01 00:00:00.005
```

But is:
```
Hello,2,2,1970-01-01 00:00:00.0,1970-01-01 00:00:00.005
Hello world,2,3,1970-01-01 00:00:00.0,1970-01-01 00:00:00.005
Hi,1,1,1970-01-01 00:00:00.0,1970-01-01 00:00:00.005
```

wuchong commented:
It's wired. I can't reproduce the wrong results in my local environment. What is the `data` in your test @twalthr ?  I'm using the following data in the batch and stream test, but the result is same. 

```scala
val data = List(
    (1L, 1, "Hi"),
    (2L, 2, "Hello"),
    (4L, 2, "Hello"),
    (8L, 3, "Hello world"),
    (6L, 3, "Hello world"))
```

cc @fhueske could your help to test this in your environment ?

twalthr commented:
@wuchong I used the data of `org.apache.flink.table.api.scala.stream.table.AggregationsITCase`.

```
  val data = List(
    (1L, 1, "Hi"),
    (2L, 2, "Hello"),
    (4L, 2, "Hello"),
    (8L, 3, "Hello world"),
    (16L, 3, "Hello world"))
```

fhueske commented:
Hi, I ran both methods with the provided data and both compute the same result on my machine.

twalthr commented:
@fhueske and I looked into this again. I seems that result depends on the batch ExecutionEnvironment. I used a regular environment while the test base uses a CollectionExecutionEnvironment. We don't know if this is a problem of your implementation or a bug in the collection environment.

wuchong commented:
Thanks @twalthr for your hint. It is a bug in my `DataSetTumbleTimeWindowAggReduceCombineFunction#combine(...)` method, that the rowtime attribute is dropped when combining. 

The collection environment will not run combine phase, but cluster environment will. That's why we can't reproduce the wrong result in test base.

BTW, do we need to activate the cluster execution mode in table IT cases ?  Currently, only collection execution mode is activated.

twalthr commented:
Thanks for fixing the bug. The collection execution mode is on purpose because we have too many IT cases at the moment. We should only test the subclasses of `FlinkRel` like `DataSetUnion`, `DataStreamAggregate` etc. with IT cases. Once this is the case we can reenable the cluster mode.

fhueske commented:
It would also be good to have a set of well selected end-to-end IT cases. But Timo is right, the previous approach of starting a mini cluster for each test case was too expensive to maintain. The collection execution is not perfect (combiners, parallelism, semantic annotations, ...), but sufficient for most cases.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetWindowAggregate.scala:
```diff
@@ -0,0 +1,244 @@
+        mappedInput.asInstanceOf[DataSet[Row]]
+          .groupBy(groupingKeys: _*)
+          // sort on time field, it's the one after grouping keys
+          .sortGroup(groupingKeys.length, Order.ASCENDING)
```
Shouldn't this be `mapReturnType.getArity - 1`? According to the docs of `AggregateUtil#createDataSetWindowPrepareMapFunction` the time field should be at the end?

wuchong replied:
Good catch ! It should be the last one in the Row.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -517,22 +657,20 @@ object AggregateUtil {
     // get all field data types of all intermediate aggregates
     val aggTypes: Seq[TypeInformation[_]] = aggregates.flatMap(_.intermediateDataType)
 
     // concat group key types and aggregation types
```
Update comment.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -591,5 +729,27 @@ object AggregateUtil {
+        if (relDataType.length == 1) {
+          relDataType.head.getIndex
+        } else {
+          throw new IllegalArgumentException()
```
I would throw a general `TableException` and give a meaningful message here and in the following 2 exceptions.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -591,5 +729,27 @@ object AggregateUtil {
+    timeField match {
+      case ResolvedFieldReference(name, resultType) =>
+        // get the RelDataType referenced by the time-field
+        val relDataType = inputType.getFieldList.filter(r => name.equals(r.getName))
```
Do we want to be case-insensitive here?

wuchong replied:
We set `Lex.JAVA` as default lexical policy which is case-sensitive. I will change this logical to adapt case-sensitive or insensitive. So that it works when user customize lexical policy.

twalthr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetTumbleCountWindowAggReduceGroupFunction.scala:
```diff
@@ -0,0 +1,93 @@
+
+    var count: Long = 0
+
+    records.foreach( (record) => {
```
Just a small comment here for future implementations. We should try to reduce Scala magic in runtime classes. I would use a good old while loop here. We don't know how many method calls and helper objects are created in this reducer.

twalthr commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/GroupWindowTest.scala:
```diff
@@ -0,0 +1,336 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.api.scala.batch.table
```
This class can be moved to `org.apache.flink.table.api.scala.batch.table`.

twalthr commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/api/scala/batch/table/AggregationsITCase.scala:
```diff
@@ -342,4 +342,85 @@ class AggregationsITCase(
     TestBaseUtils.compareResultAsText(results.asJava, expected)
   }
 
+  @Test(expected = classOf[UnsupportedOperationException])
```
I created a new package for runtime IT cases. You could move the new tests here in `org.apache.flink.table.runtime.dataset.DataSetWindowAggregateITCase`.

wuchong commented:
Hi @twalthr , thanks for your reviewing. I have addressed your comments.

twalthr commented:
Thanks for the update @wuchong. I will test it again and merge it.
